cli/boards.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	// Validate the version is valid.
+	if !supportedVersionRegex.MatchString(versionArg) {
+		return fmt.Errorf("invalid version %s. Must use semver 2.0.0 specification for versions", versionArg)
+	}
+
+	client, err := newAppClient(c)
+	if err != nil {
+		return err
+	}
+	ctx := client.c.Context
+
+	// get the org from the name or id.
+	org, err := client.getOrg(orgArg)
+	if err != nil {
+		return err
+	}
+
+	// check if the package already exists.
+	err = client.boardDefsPackageExists(ctx, org.Id, nameArg, versionArg)
+	if err != nil {
+		return err
+	}
+
+	jsonFile, err := os.Open(filepath.Clean(jsonPath))
+	if err != nil {
+		return err
+	}
+
+	if !strings.HasSuffix(jsonFile.Name(), ".json") {
+		return errors.New("The board definition file must be a .json")
+	}
+
+	_, err = client.uploadBoardDefsFile(nameArg, versionArg, org.Id, jsonFile)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(c.App.Writer, "Board definitions file was successfully uploaded!\n")
+	return nil
+}
+
+func (c *appClient) uploadBoardDefsFile(
+	name string,
+	version string,
+	orgID string,
+	jsonFile *os.File,
+) (*packagepb.CreatePackageResponse, error) {
+	if err := c.ensureLoggedIn(); err != nil {
+		return nil, err
+	}
+	ctx := c.c.Context
+
+	// Create an archive tar.gz file (required for packages).
+	file, err := CreateArchive(jsonFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating archive")
+	}
+
+	// The board defs packages are small and never expected to be larger than the upload chunk size,
+	// so we are sending in one chunk.
+	// If the file is too big, return error.
+	if file.Len() > boardUploadChunkSize {
+		return nil, fmt.Errorf("file is too large, must be under %d bytes", boardUploadChunkSize)
+	}
+
+	stream, err := c.packageClient.CreatePackage(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error starting CreatePackage stream")
+	}
+
+	stats, err := jsonFile.Stat()
+	if err != nil {
+		return nil, err
+	}
+	boardDefsFile := []*packagepb.FileInfo{{Name: name, Size: uint64(stats.Size())}}
+
+	packageInfo := &packagepb.PackageInfo{
+		OrganizationId: orgID,
+		Name:           name,
+		Version:        version,
+		Type:           packagepb.PackageType_PACKAGE_TYPE_BOARD_DEFS,
+		Files:          boardDefsFile,
+		Metadata:       nil,
+	}
+
+	// send the package requests
+	var errs error
+	if err := sendPackageRequests(stream, file, packageInfo); err != nil {
+		errs = multierr.Combine(errs, errors.Wrapf(err, "error syncing package"))
+	}
+
+	// close the stream and receive a response when finished.
+	resp, err := stream.CloseAndRecv()
+	if err != nil {
+		errs = multierr.Combine(errs, errors.Wrapf(err, "received error response while syncing package"))
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return resp, nil
+}
+
+// helper function to check if a package with this name and version already exists.
+func (c *appClient) boardDefsPackageExists(ctx context.Context, orgID, name, version string) error {
```

Comment:
I like the comment, but the name doesn't quite match this (the name implies that it doesn't check the version). Would `boardDefsVersionExists` be an intuitive name? Checking the version implies checking the name, I think, but not the other way around.